### PR TITLE
Automatically register DAGs that are used in a context manager

### PR DESCRIPTION
**Pull request by ashb**
This has caused a bit of confusion to committers (missing off the `as dag`) and is just more user friendly

This is on by default but can be disabled by passing `auto_register=False` do a DAG

### CODE REVIEW

**Comment by malthe** [comment]
@ashb there are at least two things to consider here which I stumbled upon playing around with this:

1. There's nothing special about `with` except that tasks are automatically assigned to this dag (i.e., `dag=dag`). Tying registration of DAGs to `with` is perhaps a bit confusing in this sense – also, you can use `with` on a particular dag instance multiple times!

2. What about DAGs declared in functions?

    ```python
    def test():
        with DAG():
            ...
    test()
    ```

    I don't mind but it is a little magic perhaps, that this automatically _declares_ the DAG.

I ended up instead with this pattern:
```python
dag = DAG(...)

with dag:
    ...
```
– as a recommended pattern, simply because then you can't forget to give the DAG instance a top-level variable.

And then as I mention in https://lists.apache.org/thread/xndt3zklxbzo5nlmjz10dtfm2pp4t4wq, we could instead warn the user if they'd created a DAG which isn't exposed/registered.

(The gist of that linked reply is that there are two different ways to do that.)

**Comment by kaxil** [comment]
>Tying registration of DAGs to with is perhaps a bit confusing in this sense – also, you can use with on a particular dag instance multiple times!

@malthe Is it the param name that is confusing? I do not understand I follow what you are saying. The use-case for this PR is that we don't need to needlessly do a `with DAG(..) as dag` as there is no need to have the `as dag` in it unless someone wants to actually use the `dag` variable itself

>What about DAGs declared in functions?
Isn't it the same behaviour as the one with `with DAG(..) as dag` ?

**Comment by malthe** [comment]
@kaxil I understand the use-case – what I am somewhat hesitant to vote for is the conflation of the two orthogonal concerns which are both suggested for `with` here:

1. As a scope for tasks created which automatically get the `dag` parameter set;
2. As a mechanism to automatically register a DAG with the loading mechanism.

I think you could argue then that _any_ DAG instance created during the loading of a module should be automatically registered – after all, why would a user create a DAG instance only to throw it away?

File: airflow/models/dag.py [hunk 1]
```diff
@@ -3022,24 +3029,28 @@ class DagContext:
 
     """
 
-    _context_managed_dag: Optional[DAG] = None
-    _previous_context_managed_dags: List[DAG] = []
+    _context_managed_dags: Deque[DAG] = deque()
```
Comment (uranusjr) [inline airflow/models/dag.py:3437]: Any particular reason we need a deque here? From what I can tell a plain list is enough (with the ordering reversed, i.e. we append to / pop from the end and use -1 to access).

Comment (ashb) [reply airflow/models/dag.py:3437]: No particular reason no -- mostly thinking that we only ever need to access the head value, or push a new head value, and deque is more optimized for that case.

But it hardly matters as this is not on the critical path for performance.

File: tests/models/test_dagbag.py [hunk 2]
```diff
@@ -937,8 +953,7 @@ def test_get_dag_with_dag_serialization(self):
     def test_collect_dags_from_db(self):
         """DAGs are collected from Database"""
         db.clear_db_dags()
-        example_dags_folder = airflow.example_dags.__path__[0]
-        dagbag = DagBag(example_dags_folder)
+        dagbag = DagBag(str(example_dags_folder))
```
Comment (uranusjr) [inline tests/models/test_dagbag.py:963]: Maybe we should just allow `DagBag` to take `Path`, but this doesn’t need to be done now.

**Review by uranusjr — APPROVED** [review]

File: tests/models/test_dagbag.py (hunk 2 above)
Comment (ashb) [reply tests/models/test_dagbag.py:963]: Yeah, that was my thought -- it was orthogonal to this PR

File: newsfragments/23592.significant.rst [hunk 3]
```diff
@@ -0,0 +1,24 @@
+DAGS used in a context manager no longer need to be assigned to a module variable
+
+Previously you had do assign a DAG to a module-level variable in order for Airflow to pick it up. For example this
```
Comment (feluelle) [inline newsfragments/23592.significant.rst:3]:
```suggestion
DAGs used in a context manager no longer need to be assigned to a module variable.

Previously you had to assign a DAG to a module-level variable in order for Airflow to pick it up. For example this
```

File: newsfragments/23592.significant.rst [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+can become
+
+.. code-block::
+
+   with DAG(dag_id="example"):
+      ...
+
+
+If you want to disable the behaviour for any reason then set ``auto_register=False`` on the dag::
```
Comment (feluelle) [inline newsfragments/23592.significant.rst:34]:
```suggestion
If you want to disable the behaviour for any reason then set ``auto_register=False`` on the dag
```

File: airflow/models/dag.py [hunk 5]
```diff
@@ -312,6 +316,7 @@ class DAG(LoggingMixin):
         ``Environment`` is used to render templates as string values.
     :param tags: List of tags to help filtering DAGs in the UI.
     :param schedule_on: List of upstream datasets if for use in triggering DAG runs.
+    :param auto_register: Automatically register this DAG when it is used in a ``with`` block
```
Comment (feluelle) [inline airflow/models/dag.py:336]: If this is only applicable to context manager calls I think this is not the correct place to pass it or? Wouldn't it make more sense to define this on config level for all DAGs using context manager i.e. in the config?

Comment (ashb) [reply airflow/models/dag.py:336]: It's only relevant to context manager yes, but this is the only place you can add a flag to change the behavoiur of _one_ dag.

I think we could add a new config flag (default to True) that globally sets the auto_register behaviour, but I think we still need the way to change the behaviour for a single dag.

The other option I can think of is this:

```python
with DAG(dag_id=...).with_autoregister(False):
    ...
```

I don't have a strong preference either way

Comment (potiuk) [reply airflow/models/dag.py:336]: Yeah. I like it the way it is now. with X.with_something() does not look good, and the current way is more explicit.

**Review by potiuk — APPROVED** [review]
LGTM

**Comment by ashb** [comment]
I still need to update the docs/tutorial etc to reflect this.

File: airflow/models/dag.py (hunk 5 above)
Comment (feluelle) [reply airflow/models/dag.py:336]: Yeah if we only have these two options, I would also keep it like it is. 👍

**Comment by ashb** [comment]
Reminder to self: finish this before 2.4

File: airflow/example_dags/tutorial.py [hunk 6]
```diff
@@ -97,7 +97,7 @@
     """
     )
 
-    dag.doc_md = __doc__  # providing that you have a docstring at the beginning of the DAG
+    dag.doc_md = __doc__  # providing that you have a docstring at the beginning of the DAG; OR
```
Comment (uranusjr) [inline airflow/example_dags/tutorial.py:100]:
```suggestion
    dag.doc_md = __doc__  # providing that you have a docstring at the beginning of the DAG
```

accidental?

Comment (ashb) [reply airflow/example_dags/tutorial.py:100]: No, intentional but probably shouldn't be in this PR. (The tutorial is showing multiple ways of setting `dag.doc` -- you wouldn't want to do all of them.)

Comment (uranusjr) [reply airflow/example_dags/tutorial.py:100]: Oh, this needs to be read with a next lines as one sentence. Got it now.

**Review by potiuk — APPROVED** [review]

File: airflow/models/dagbag.py [hunk 7]
```diff
@@ -306,6 +314,8 @@ def _load_modules_from_file(self, filepath, safe_mode):
         if mod_name in sys.modules:
             del sys.modules[mod_name]
 
+        DagContext.current_autoregister_module_name = mod_name
```
Comment (uranusjr) [inline airflow/models/dagbag.py:317]: This in-place operation feels a bit weird to me… Perhaps a context manager would be easier to understand and maintain? Something like

```python
with DagContext.enable_autoregister(mod_name):
    parse(...)
```

This probably needs some refactoring to work though.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 8]
```diff
@@ -133,13 +134,16 @@ As long as a ``DAG`` object in ``globals()`` is created, Airflow will load it.
 
             print_message(config["message"])
 
-        globals()[dag_id] = dynamic_generated_dag()
+        dynamic_generated_dag()
 
 The code below will generate a DAG for each config: ``dynamic_generated_dag_config1`` and ``dynamic_generated_dag_config2``.
```
Comment (jedcunningham) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst:139]:
```suggestion
The code above will generate a DAG for each config: ``dynamic_generated_dag_config1`` and ``dynamic_generated_dag_config2``.
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 9]
```diff
@@ -107,8 +107,9 @@ the meta-data file in your DAG easily. The location of the file to read can be f
     # Configuration dict is available here
 
 
-Dynamic DAGs with ``globals()``
-...............................
+Registering dynamic DAGs
+........................
+
 You can dynamically generate DAGs by working with ``globals()``.
 As long as a ``DAG`` object in ``globals()`` is created, Airflow will load it.
```
Comment (jedcunningham) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]: Maybe something like this instead?

```suggestion
You can dynamically generate DAGs when using the ``@dag`` decorator or the ``with DAG(..)`` context manager.
Airflow will automatically register them.
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 10]
```diff
@@ -133,13 +134,16 @@ As long as a ``DAG`` object in ``globals()`` is created, Airflow will load it.
 
             print_message(config["message"])
 
-        globals()[dag_id] = dynamic_generated_dag()
+        dynamic_generated_dag()
 
 The code below will generate a DAG for each config: ``dynamic_generated_dag_config1`` and ``dynamic_generated_dag_config2``.
 Each of them can run separately with related configuration
```
Comment (jedcunningham) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
Each of them can run separately with related configuration.

If you do not wish to have DAGs auto-registered, you can disable the behavior by setting ``auto_register=False`` on your DAG.
```

File: newsfragments/23592.significant.rst [hunk 11]
```diff
@@ -0,0 +1,24 @@
+DAGS used in a context manager no longer need to be assigned to a module variable
```
Comment (jedcunningham) [inline newsfragments/23592.significant.rst:1]: This should also cover the `@dag` decorator, no?

File: airflow/models/dagbag.py (hunk 7 above)
Comment (ashb) [reply airflow/models/dagbag.py:317]: Yeah, I agree, This is essentially used as a global variable.

It's not very easy to refactor given it's use across three or so different functions, so a context manager would be tricky. (I.e. I can't see an easy refactor to make it work right now)

**Review by jedcunningham — APPROVED** [review]

**Pull request merged**